orca-qos/src/main/kotlin/com/netflix/spinnaker/orca/qos/ExecutionBufferActuator.kt

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2018 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.orca.qos
+
+import com.netflix.spinnaker.orca.ExecutionStatus.BUFFERED
+import com.netflix.spinnaker.orca.annotations.Sync
+import com.netflix.spinnaker.orca.events.BeforeInitialExecutionPersist
+import com.netflix.spinnaker.orca.pipeline.model.Execution
+import com.netflix.spinnaker.orca.qos.BufferAction.*
+import com.netflix.spinnaker.orca.qos.BufferState.ACTIVE
+import net.logstash.logback.argument.StructuredArguments.value
+import org.slf4j.LoggerFactory
+import org.springframework.context.event.EventListener
+import org.springframework.stereotype.Component
+
+/**
+ * Determines if an execution should be buffered.
+ */
+@Component
+class ExecutionBufferActuator(
+  private val bufferStateSupplier: BufferStateSupplier,
+  policies: List<BufferPolicy>
+) {
+
+  private val log = LoggerFactory.getLogger(ExecutionBufferActuator::class.java)
+
+  private val orderedPolicies = policies.sortedByDescending { it.order }.toList()
+
+  @Sync
+  @EventListener(BeforeInitialExecutionPersist::class)
+  fun beforeInitialPersist(event: BeforeInitialExecutionPersist) {
+    if (bufferStateSupplier.get() == ACTIVE) {
+      val execution = event.execution
+      withActionDecision(execution) {
+        when (it.action) {
+          BUFFER -> {
+            log.warn("Buffering execution: {}, reason: ${it.reason}", value("executionId", execution.id))
+            execution.status = BUFFERED
+          }
+          ENQUEUE -> {
+            log.debug("Enqueuing execution: {}, reason: ${it.reason}", value("executionId", execution.id))
+          }
+          ABSTAIN -> {
+            log.debug("Abstaining from quality of service, enqueuing execution: {}, reason: ${it.reason}", value("executionId", execution.id))
+          }
+        }
+      }
+    }
+  }
+
+  fun withActionDecision(execution: Execution, fn: (BufferResult) -> Unit) {
+    orderedPolicies
+      .map { it.apply(execution) }
+      .let { policies ->
+        // If any policies have flagged themselves to force an action, the first force is used. Otherwise, the last
+        // policy's action will be used.
+        // TODO rz - In the future, better heuristics around all policies informing the decision may need to be
+        // implemented.
+        try {
+          return@let policies.firstOrNull { it.force } ?: policies.last()
```

Comment:
Why does the last policy win and how is ordering determined? It's not clear to me what behavior should be expected if i.e. there are 3 active policies where 1 returns BUFFER and 2 ENQUEUE, but none set force==true. Still wrapping my head around the full PR, but I have an initial naive preference for considering BUFFER=false, ENQUEUE=true, ABSTAIN=true where an execution will be enqueued `if (policy1 && policy2 ... && policyN) == true`.

Reply:
This decision logic is by no means ideal (I even have a TODO right above saying that this needs to be done better). I think requiring all to ENQUEUE is probably the easiest strategy to reason about. Do like.

Reply:
I feel like `ENQUEUE` winning is probably right. Err on the side of caution.

Reply:
Although you could also say erring on the side of caution is to `BUFFER` the execution. If the system is under load, I would want the system to BUFFER by default. But yes, we should simplify.

Reply:
D'oh. That's what I meant. Sorry.